[kwuunnn] iP

kwuunnn opened the pull request:
# KwunTalk
> A Chatbot that helps manage tasks.

Tick the boxes if you:

- [ ] Always forget what deadlines or events you have
- [ ] Have low productivity
- [ ] Want a simple and ~~fast~~ **_SUPER FAST_** way to manage tasks

If you tick any of the boxes, then **KwunTalk** is _definitely_ for you! :+1:

## Features

- Supports the creation of Todo, Deadline and Event tasks.
- Allows marking of task as done or undone and deletion of tasks.
- Automatically stores task list when exit.

## Launch
**To run the Chatbot:**

1. Go to the [releases](https://github.com/kwuunnn/ip/releases/tag/A-Jar)
2. Download the `duke.jar` file
3. Move the file into an empty folder
4. Open a command window in that folder
5. Run `java -jar duke.jar` (in the same folder as the jar file)

---
> [!NOTE]
> If you are a Java programmer, you can use it to practice Java too. 
>
> Here's the `main` method:
```java
public static void main(String[] args) {
    new Duke("./data/duke.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

kwuunnn committed:
feat: Implement skeletal version

Duke Chatbot template with no features

Let's
* Rename Chatbot from Duke to KwunBot
* Implement greeting
* Implement bye statement

kwuunnn committed:
feat: Implement actions on user input

Skeletal version with simple greeting and bye

Let's
* Echo commands entered by user
* Exit with bye statement when user inputs 'bye'

kwuunnn committed:
feat: Implement list

Chatbot can echo user inputs

Let's
* Add the ability to store tasks in list
* Implement 'list' command to show current list of tasks stored

kwuunnn committed:
feat: Implement checklist

Chatbot can store tasks in list

Let's
* Add Task class to represent a task in the list
* Add the ability to mark task as done or undone
* Allow user to see checklist

kwuunnn committed:
feat: Implement 3 specific types of task: ToDos, Deadlines, Events

Chatbot handles all tasks the same way.

Provide specific support for the 3 different types of tasks.
Creating 3 subclasses for each type of task allows us to use
polymorphism when dealing with tasks.

Let's
* Add 'Todo', 'Deadline', 'Event' classes to represent the respective task
* Pull up behaviours common to all tasks classes into the parent 'Task' class
* Make 'Task' class an abstract class

kwuunnn committed:
fix: No line found by Scanner

An infinite while loop to scan user inputs

NoSuchElementException: No line found at Scanner.nextLine

Let's replace while true to while Scanner hasNextLine()

To check if there is a next line before scanning.

kwuunnn committed:
feat: Implement Automated Text UI Testing

Manually testing Chatbot.

Use I/O redirection technique to semi-automate testing.

Let's
* Create text-ui-test folder
* Add runtest.bat (Windows)
* Add input.txt and EXPECTED.txt for testing

kwuunnn committed:
feat: Implement error handling

Chatbot did not handle errors and threw exceptions.

Create Exceptions classes to handle errors.

Let's
* Create abstract class DukeException
* Create subclasses to handle specific errors
* Handle errors by catching them and outputing appropriate response

kwuunnn committed:
feat: Implement delete task

Chatbot not able to delete task.

Use ArrayList to store tasks and allow task deletion.

Let's
* Create use ArrayList to store tasks
* Implement delete command
* Handle possible errors with delete command

kwuunnn committed:
fix: Make commands to lowercase

Uppercase input would produce wrong response.

Let's use toLowerCase() to make the command lowercased.

kwuunnn committed:
fix: Text UI Testing input.txt and EXPECTED.TXT

Expected result was incorrect due to changes made in Chatbot.

Let's edit EXPECTED.TXT and change input.txt to incorporate new features.

kwuunnn committed:
feat: Use Enumerations

Command types are constants so using Enums makes sense.

Let's create CommandType enum and use switch statement to check for
corresponding values.

kwuunnn committed:
fix: Fix 'bye' command

Chatbot was not exiting upon 'bye' command

Let's implement a shouldBreak variable and an if statement to break out of while loop.

kwuunnn committed:
feat: Implement task saving

Chatbot would lose data upon restart.

Create tasks file to store data and load data when chatbot starts up.

Let's
* Use File access to read file
* Use File access to write to file
* Handle errors regarding file access

kwuunnn committed:
Merge branch 'branch-Level-7'

kwuunnn committed:
feat: Use Date and Time

Chatbot did not understand dates and times.

Use java.time.LocalDateTime.

Let's
* Accept datetimes in dd/MM/yyyy HHmm format
* Print dates in a different format
* Handle erros pertaining to wrong date format inputted

kwuunnn committed:
Merge branch 'branch-Level-8'

kwuunnn committed:
feat: Use More OOP

Code was repetitive and mostly in one file.

Extract out closely related code.

Let's
* Extract Ui, Storage, Parser, TaskList classes
* Created abstract Command classes and more *Command classes
* Handled more exceptions

kwuunnn committed:
refactor: Organize into packages

Code was messy.

Divide classes into packages.

Let's create duke duke.task, duke.command, duke.exception packages.

kwuunnn committed:
Merge branch 'add-gradle-support' of https://github.com/kwuunnn/ip

kwuunnn committed:
feat: Automate project builds using Gradle

Let's
* Merge add-gradle-support branch from origin to master branch
* Edit build.gradle and set mainClass to 'duke.Duke'

kwuunnn committed:
feat: Implement Unit testing

Let's use JUnit tests to test the behaviour of the code.

kwuunnn committed:
docs: Add JavaDoc comments

Let's add JavaDoc comments to all non-private classes/methods.

kwuunnn committed:
feat: Implement Find command

Give users a way to find a task by searching for a keyword.

Let's
* Create FindCommand class
* Create Ui method to show find command interface
* Include Find command in Parser

kwuunnn committed:
Merge branch 'branch-Level-9'

kwuunnn committed:
Merge branch 'master' into branch-A-JavaDoc

kwuunnn committed:
docs: Add JavaDoc comments

Let's add JavaDoc comments to all non-private classes/methods.

kwuunnn committed:
Merge branch 'branch-A-JavaDoc'

kwuunnn committed:
docs: Edit JavaDoc comments

kwuunnn committed:
docs: Edit Javadocs

kwuunnn committed:
docs: Edit Javadocs

kwuunnn committed:
perf: Implement checkstyle

Use checkStyle to detect coding style violations.

kwuunnn committed:
Merge branch 'branch-A-CheckStyle'

YowSiaoKang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,112 @@
+        return tasks;
+    }
+
+    private void createTaskFile() throws TasksFileException {
```
Do take note of the naming here to maintain consistency.
```suggestion
    private void createTasksFile() throws TasksFileException {
```

YowSiaoKang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,63 @@
+            case EVENT:
+                return new AddEventCommand(parts[1]);
+            case FIND:
+                return new FindCommand(parts[1]);
```
I like how the commands are named consistently.

YowSiaoKang commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,90 @@
+     * @param task Task to be added.
+     */
+    public void add(Task task) {
+        this.taskList.add(task);
```
Perhaps you can specify what is added in the function name.

YowSiaoKang commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,53 @@
+    /**
+     * Executes the delete command.
+     *
+      * @param taskList List of tasks.
```
There seems to be an extra space here!

YowSiaoKang commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,64 @@
+            throws InvalidArgumentException, InvalidDateTimeFormatException {
+
+        try {
+            String[] splitDate = this.arguments.split(" /by ", 2);
```
I feel the naming here can be improved.
```suggestion
            String[] argumentParts = this.arguments.split(" /by ", 2);
```

YowSiaoKang reviewed (commented):
Looks good to me, just some issues with naming. Good Job!

kwuunnn committed:
feat: Implement GUI

Use JavaFx technology to implement GUI.

kwuunnn committed:
Merge branch 'branch-Level-10'

kwuunnn committed:
feat: Use Varargs

Use Varargs for TaskLists constructor.

kwuunnn committed:
fix: Shorten length of divider line to fit nicely in the main window in GUI.

kwuunnn committed:
Merge branch 'branch-A-Varargs'

g-tejas commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,88 @@
+     * @param tasks Varargs of tasks added to the tasklist.
+     */
+    public TaskList(Task... tasks) {
+        this.taskList.addAll(List.of(tasks));
```
I like the usage of varargs here

g-tejas commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,88 @@
+     * @return Filtered list of tasks.
+     */
+    public TaskList filter(String keyword) {
+        Stream<Task> filteredTasks = this.taskList.stream()
```
I like the usage of Java streams to do keyword searching

g-tejas commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,149 @@
+            return DIVIDER + "There are no matching tasks in your list.\n" + DIVIDER;
+
+        } else {
+            StringBuilder sb = new StringBuilder();
```
Perhaps you can initialize the string builder with the leading text, like
`Stringbuilder sb = new StringBuilder("Here are the ...");`

g-tejas commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,46 @@
+     * @return True for Bye commands.
+     */
+    @Override
+    public boolean isExit() {
+        return true;
+    }
```
Is this needed? Seems like a lot of overhead for one type of command, could be handled at the main loop itself?

g-tejas reviewed (commented):
Overall LGTM, just some nitpicks but nothing serious.

kwuunnn committed:
Edit test files

TaskListTest.java file raised exception.

TaskList constructor uses Varargs instead of ArrayList.

Let's use toArray() to convert ArrayList to Task[].

kwuunnn committed:
Edit build.gradle

Gradle run not running the JavaFx GUI.

MainClass was set to duke.Duke.

Let's set the mainClass to duke.Launcher.

JavaFx GUI can be launched.

kwuunnn committed:
Add assertions

Invariants were assumed and possible errors were not handled.

Let's use assertions to verify assumptions about invariants in some methods and classes.

kwuunnn committed:
Improve code quality

Method and variable names were not explanatory.

Let's change method names to using verbs.

kwuunnn committed:
Merge pull request #1 from kwuunnn/branch-A-Assertions

Add assertions

kwuunnn committed:
Merge branch 'master' into branch-A-CodeQuality

kwuunnn committed:
Merge pull request #2 from kwuunnn/branch-A-CodeQuality

Improve code quality

kwuunnn committed:
Implement detection of duplicate tasks

Added task could be a duplicate in the database.

Let's check for duplicates and raise an error if duplicate task is found.

kwuunnn committed:
Merge pull request #3 from kwuunnn/branch-C-DetectDuplicates

Implement detection of duplicate tasks

kwuunnn committed:
Refactor Duke Chatbot to KwunTalk

Packages, Classes and Methods were named according to Duke.
Standard images were used for the chatbot.

Let's
* Rename everything from Duke to KwunTalk.
* Add personalised images to be used in Chatbot.

kwuunnn committed:
Improve code quality and standard

Let's fix checkstyle errors.

kwuunnn committed:
Add a representative screenshot of product

Let's add Ui.png to the docs folder

kwuunnn committed:
Edit docs\README.md

Let's add a User Guide for KwunTalk

kwuunnn committed:
Edit docs\README.md

Let's edit  User Guide for KwunTalk

kwuunnn committed:
Edit docs\README.md

Let's edit User Guide for KwunTalk

kwuunnn committed:
Edit docs\README.md

Let's edit User Guide for KwunTalk

kwuunnn committed:
Edit docs\README.md

Let's edit User Guide for KwunTalk

kwuunnn committed:
Make GUI Window close upon exit of program

The 'bye' command did not close the window.

Let's
* Create a boolean variable to check if program should exit.
* Close the GUI window when bye command executed.

kwuunnn committed:
Make tasks list save automatically everytime the task changes

Task list was only saved upon the 'bye' command.

Let's call the save method whenever task list changes:
* adding todo.
* adding deadline.
* adding event.
* marking task as done.
* marking task as not done.
* deleting task.
